core/src/polling_monitor/ipfs_service.rs

```diff
@@ -0,0 +1,97 @@
+use anyhow::{anyhow, Error};
+use bytes::Bytes;
+use cid::Cid;
+use futures::{Future, FutureExt};
+use graph::{cheap_clone::CheapClone, ipfs_client::IpfsClient, tokio::sync::Semaphore};
+use std::{pin::Pin, sync::Arc, task::Poll, time::Duration};
+use tower::Service;
+
+const CLOUDFLARE_TIMEOUT: u16 = 524;
+const GATEWAY_TIMEOUT: u16 = 504;
+
+#[derive(Clone)]
+pub struct IpfsService {
+    client: IpfsClient,
+    max_file_size: u64,
+    timeout: Duration,
+    concurrency_limiter: Arc<Semaphore>,
+}
+
+impl CheapClone for IpfsService {
+    fn cheap_clone(&self) -> Self {
+        Self {
+            client: self.client.cheap_clone(),
+            max_file_size: self.max_file_size,
+            timeout: self.timeout,
+            concurrency_limiter: self.concurrency_limiter.cheap_clone(),
+        }
+    }
+}
+
+impl IpfsService {
+    #[allow(dead_code)]
+    pub fn new(
+        client: IpfsClient,
+        max_file_size: u64,
+        timeout: Duration,
+        concurrency_limit: u16,
+    ) -> Self {
+        Self {
+            client,
+            max_file_size,
+            timeout,
+            concurrency_limiter: Arc::new(Semaphore::new(concurrency_limit as usize)),
+        }
+    }
+
+    async fn call(&self, cid: Cid) -> Result<Option<Bytes>, Error> {
+        let cid_str = cid.to_string();
+        let stat = match self.client.object_stat(&cid_str, self.timeout).await {
+            Ok(stat) => stat,
+            Err(e) => match e.status().map(|e| e.as_u16()) {
+                Some(GATEWAY_TIMEOUT) | Some(CLOUDFLARE_TIMEOUT) => return Ok(None),
+                _ if e.is_timeout() => return Ok(None),
```

Comment:
Could these two `match` arms be together? 🤔

Reply:
This was the best I came up with but open to suggestions.

Reply:
That wouldn't work, the `if` clause applies to the entire pattern, I just tried it on the [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7290983d0e09110ef5710411c85958b4).